docs(publishReplay): Add description and examples

okeefem2 opened the pull request:
**Description:** Add documentation and examples for publishReplay operator.

**Related issue (if exists):** #4802 

Hello! This is my first contribution to rxjs, I use this library a lot, but I had to do some learning for this operator and it is a working knowledge more than theoretical that I have, so if any of the descriptions are wrong or don't quite make sense please let me know and I will work to make them better.

benlesh commented on src/internal/operators/publishReplay.ts:
````diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * Make source$ hot by applying publishReplay operator, and cache 1 value in a ReplaySubject
+ * for any observers that subscribe after the source begins emitting.
+ * ```ts
+ * import { of, zip, interval, merge } from "rxjs";
````
Some of the imports here are unused. Also, use single quotes, please :)

benlesh commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * import { map, publishReplay, tap, refCount } from "rxjs/operators";
+ *
+ * const source$ = zip(interval(1000), of(1, 2, 3, 4, 5, 6, 7, 8, 9)).pipe(
+ *  map(values => values[1]),
```
formatting/indentation looks a little off.

benlesh commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ *
+ * // publishReplay can take an optional second argument windowTime to delay the emission of the cached values
+ * // to observers that subscribe after the source begins.
+ * import { of, zip, interval, merge } from "rxjs";
```
Same as above.

benlesh commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * import { of, zip, interval, merge } from "rxjs";
+ * import { map, publishReplay, tap, refCount } from "rxjs/operators";
+ *
+ * const source$ = zip(interval(1000), of(1, 2, 3, 4, 5, 6, 7, 8, 9)).pipe(
```
Just use `interval(1000)`, the `zip` and `of` are a little superfluous here, we want to keep the examples as simple as possible.

benlesh commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * const source$ = zip(interval(1000), of(1, 2, 3, 4, 5, 6, 7, 8, 9)).pipe(
+ *  map(values => values[1]),
+ *    publishReplay(1),
+ *    refCount()
```
Perhaps show an example of how to cast as `ConnectableObservable` and call `connect()` as well?

benlesh commented on src/internal/operators/publishReplay.ts:
````diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * // Stream 1: 9
+ * // Stream 2: 9
+ * ```
+ *
````
I love the work here so far. Ironically, though, you haven't touched on the use case I'd most like to see people use that doesn't involve `refCount` or `ConnectableObservable` at all. You can multicast with a selection function:

```ts
of(1, 2, 3, 4).pipe(
  publishReplay(3, 3000, shared$ => merge(
    shared$.pipe(delay(1000), map(v => `one: ${v}`)),
    shared$.pipe(delay(300), map(v => `two ${v}`)),
  ))
  .subscribe({
    next: x => console.log(x),
    complete: () => console.log('done'),
  })

// Output
// one 1
// one 2
// one 3
// two 1
// one 4
// two 2
// two 3
// two 4
// done
```

okeefem2 replied:
A good point! Admittedly I have not used it that way before and I am glad to have learned it. Thanks a bunch for all the suggestions and the example, and good eye on the code style issues. I will work on these changes tonight and get them up for your review when you have the time! @benlesh

okeefem2 replied:
Okay I know it's been awhile this, I apologize! I have been in the middle of some life changes. But finally got back around to this. I think I am missing a piece of this somewhere...

So I took the example you had above and played around with it a bit to try to get a better understanding and I came up with this example:

`
import { merge, interval } from 'rxjs';
import { publishReplay, map, delay, takeWhile, publish } from 'rxjs/operators';

const source$ = interval(1000).pipe(
  takeWhile(x => x < 5),
  publishReplay(
    1,
    0,
    shared$ => merge(
      shared$.pipe(delay(1000), map(v => `Stream 1: ${v}`)),
      shared$.pipe(delay(500), map(v => `Stream 2: ${v}`)),
    ),
  ),
 );

 source$.subscribe({
  next: x => console.log(`Sub 1: ${x}`),
  complete: () => console.log('Sub 1 done'),
 });

setTimeout(() => {
   source$.subscribe({
    next: x => console.log(`Sub 2: ${x}`),
    complete: () => console.log('Sub 2 done'),
  });
}, 3000);

// Sub 1: Stream 2: 0
// Sub 1: Stream 1: 0
// Sub 1: Stream 2: 1
// Sub 1: Stream 1: 1
// Sub 1: Stream 2: 2
// Sub 1: Stream 1: 2
// Sub 1: Stream 2: 3
// Sub 1: Stream 2: 0
// Sub 2: Stream 2: 3
// Sub 2: Stream 2: 0
// Sub 1: Stream 1: 3
// Sub 2: Stream 1: 3
// Sub 2: Stream 1: 0
// Sub 1: Stream 1: 0
// Sub 1: Stream 2: 4
// Sub 1: Stream 2: 1
// Sub 2: Stream 2: 4
// Sub 2: Stream 2: 1
// Sub 1: Stream 1: 4
// Sub 1: Stream 1: 1
// Sub 2: Stream 1: 4
// Sub 2: Stream 1: 1
// Sub 1 done
// Sub 2 done
`

This is different from what I expected to see, so I was hoping for some clarification!
The second subscription picks up the replayed emissions from the selector as I would expect (`Sub 2: Stream 2: 3` and `Sub 2: Stream 1: 3` emit when the second subscription is created) but that subscription also seems to restart the whole stream over again and BOTH subscriptions restart back at 0 again after the second subscription begins (and they both finish out the original stream to 4). What part of this causes the streams to restart like that?

benlesh commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
 /* tslint:enable:max-line-length */
 
+/**
+ * Returns a ConnectableObservable, which is a variety of Observable that waits until its connect method is called
```
```suggestion
 * Returns a [@link ConnectableObservable}, which is a variety of observable that waits until its connect method is called
```

benlesh replied:
This isn't _completely_ accurate. 

This "multicasts over a {@link ReplaySubject}", and that's the best high-level first sentence. 

Because it doesn't _always_ return a `ConnectableObservable`. If you pass a selector function to it (I talk about this below) it will return a normal `Observable`.

JWO719 commented:
Hi @okeefem2,
thank you so much for this amazing contribution. Ben definetly made some valid points here. If you need any kind of help applying his requested changes, just let me know :)

okeefem2 commented:
Thanks so much @JWO719! I didn't get through as much as I wanted last night, but will work on it some more tonight and this weekend. If I run into any issues I will definitely reach out.

benlesh commented:
@okeefem2 did you have a chance to go through my requested changes?

JWO719 commented:
pinging @okeefem2 . Do you need any kind of help? I would really love to merge these changes, it's a great improvement to the current docs

okeefem2 commented:
@JWO719 Hey! Sorry I completely forgot about this!! I changed jobs and moved between starting this and now and life got really crazy. Taking a look tonight and will let you know.

okeefem2 commented:
Okay pushed my changes. Admittedly I am a little tired so I will plan on checking this out in the AM to make sure I didn't miss anything, but feel free to check it out and give more feedback.

I think the reworked examples cover more of the cases that @benlesh suggested and make a lot more sense at least to me... Thanks for your patience everyone!

joshribakoff commented on src/internal/operators/publishReplay.ts:
```diff
@@ -9,6 +9,120 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
+ * Can also return a {@link Observable} when a selector function is used as the third argument,
+ * this is generally the recommended way of using publish and publishReplay
+ * as it allows you to immeditately subscribe without connecting, and makes further composition easier.
+ * Check out <a href="https://blog.angularindepth.com/rxjs-calling-publish-with-a-selector-3ab48f052a4b">this post</a>
```
link not loading for me, invalid SSL

niklas-wortmann commented:
@okeefem2 Sorry that I didn't come back to this yet. If you could rebase the pr and mark the requested changes by Ben as resolved this would be highly appreciated. I would love to have those elaborations in the new v7 docs.

niklas-wortmann commented:
closing this for now. @okeefem2 feel free to reopen if you want to
